src/VisualStudio/Core/Impl/SolutionExplorer/DiagnosticItem/BaseDiagnosticAndGeneratorItemSource.cs

```diff
@@ -7,187 +7,208 @@
-                var newAnalyzerConfigOptions = project.GetAnalyzerConfigOptions();
-
-                if (newGeneralDiagnosticOption != _generalDiagnosticOption ||
-                    !object.ReferenceEquals(newSpecificDiagnosticOptions, _specificDiagnosticOptions) ||
-                    !object.ReferenceEquals(newAnalyzerConfigOptions?.TreeOptions, _analyzerConfigOptions?.TreeOptions) ||
-                    !object.ReferenceEquals(newAnalyzerConfigOptions?.ConfigOptions, _analyzerConfigOptions?.ConfigOptions))
-                {
-                    _generalDiagnosticOption = newGeneralDiagnosticOption;
-                    _specificDiagnosticOptions = newSpecificDiagnosticOptions;
-                    _analyzerConfigOptions = newAnalyzerConfigOptions;
-
-                    Contract.ThrowIfNull(_items, "We only subscribe to events after we create the items, so this should not be null.");
-
-                    foreach (var item in _items.OfType<DiagnosticItem>())
-                    {
-                        var effectiveSeverity = item.Descriptor.GetEffectiveSeverity(project.CompilationOptions, newAnalyzerConfigOptions?.ConfigOptions, newAnalyzerConfigOptions?.TreeOptions);
-                        item.UpdateEffectiveSeverity(effectiveSeverity);
-                    }
-                }
-            }
+            var result = await client.TryInvokeAsync<IRemoteSourceGenerationService, ImmutableArray<SourceGeneratorIdentity>>(
+                project,
+                (service, solutionChecksum, cancellationToken) => service.GetSourceGeneratorIdentitiesAsync(
+                    solutionChecksum, project.Id, analyzerFileReference.FullPath, cancellationToken),
+                cancellationToken).ConfigureAwait(false);
+
+            // If the call fails, the OOP substrate will have already reported an error
+            if (!result.HasValue)
+                return [];
+
+            return result.Value;
+        }
+    }
+
+    private void OnWorkspaceChanged(object sender, WorkspaceChangeEventArgs e)
+    {
+        switch (e.Kind)
+        {
+            // Solution is going away or being reloaded. The work queue will detect this and clean up accordingly.
+            case WorkspaceChangeKind.SolutionCleared:
+            case WorkspaceChangeKind.SolutionReloaded:
+            case WorkspaceChangeKind.SolutionRemoved:
+            // The project itself is being removed.  The work queue will detect this and clean up accordingly.
+            case WorkspaceChangeKind.ProjectRemoved:
+            case WorkspaceChangeKind.ProjectChanged:
+            // Could change the severity of an analyzer.
+            case WorkspaceChangeKind.AnalyzerConfigDocumentAdded:
+            case WorkspaceChangeKind.AnalyzerConfigDocumentChanged:
+            case WorkspaceChangeKind.AnalyzerConfigDocumentReloaded:
+            case WorkspaceChangeKind.AnalyzerConfigDocumentRemoved:
+                _workQueue.AddWork();
+                break;
```

Comment:
all of these events should be rare.  So we'll actually only do updates in a tiny number of workspace events.